internal/builders/docker/pkg/common.go

```diff
@@ -0,0 +1,98 @@
+// Based on BuildDefinition in https://github.com/slsa-framework/slsa/pull/525.
+type BuildDefinition struct {
+	// BuildType indicates how to unambiguously interpret this BuildDefinition.
+	BuildType string `json:"buildType"`
+	// The set of top-level external inputs to the build. This SHOULD contain all
+	// the information necessary and sufficient to initialize the build and begin
+	// execution. "Top-level" means that it is not derived from another input.
+	//
+	// REQUIRED for SLSA Build L1.
+	ExternalParameters ParameterCollection `json:"externalParameters"`
+	// Parameters of the build environment that were provided by the `builder` and
+	// not under external control. The primary intention of this field is for
+	// debugging, incident response, and vulnerability management. The values here
+	// MAY be necessary for reproducing the build.
+	//
+	// OPTIONAL.
+	SystemParameters ParameterCollection `json:"systemParameters"`
+
+	// Resolved dependencies needed at build time.
+	//
+	// OPTIONAL.
+	ResolvedDependencies []ArtifactReference `json:"resolvedDependencies"`
+}
+
+// ParameterCollection is a collection of parameters that appear in a build definition.
+type ParameterCollection struct {
+	// References to the top-level, independent input artifacts to the build. In
+	// many cases, this is a singular "source" artifact to be built.
+	//
+	// The key is a name whose interpretation depends on `buildType`. If there is
+	// only one input, it SHOULD be named "source".
+	Artifacts map[string]ArtifactReference `json:"artifacts"`
+
+	// Other parameters that are not artifact references. Like `artifacts`, the
+	// key is a name whose interpretation depends on `buildType`.
+	Values map[string]string `json:"values"`
+}
+
+// ArtifactReference contains details about an artifact.
+type ArtifactReference struct {
+	// [URI] describing where this artifact came from. When possible, this SHOULD
+	// be a universal and stable identifier, such as a source location or Package
+	// URL ([purl]).
+	//
+	// Example: `pkg:pypi/pyyaml@6.0`
+	//
+	// REQUIRED.
+	URI string `json:"uri"`
+
+	// A map of cryptographic digests for the contents of this artifact.
+	// The key indicates the cryptographic algorithm used for computing the digest.
+	//
+	// REQUIRED.
+	Digest map[string]string `json:"digest"`
+
+	// The name for this artifact local to the build.
+	//
+	// Example: `PyYAML-6.0.tar.gz`
+	//
+	// OPTIONAL.
+	LocalName string `json:"localName"`
+
+	// [URI] identifying the location that this artifact was downloaded from, if
+	// different and not derivable from `uri`.
+	//
+	// Example: `https://files.pythonhosted.org/packages/36/2b/61d51a2c4f25ef062ae3f74576b01638bebad5e045f747ff12643df63844/PyYAML-6.0.tar.gz`
+	//
+	// OPTIONAL.
+	DownloadLocation string `json:"downloadLocation"`
+
+	// [Media Type] (aka MIME type) of this artifact.
+	//
+	// OPTIONAL.
+	MediaType string `json:"mediaType"`
```

Comment:
for the optional fields, can you please add `omitempty`?  like
```suggestion
	MediaType string `json:"mediaType,omitempty"`
```

Reply:
Good catch, thanks. I've add `omitempty`, perhaps rather generously. Please take another look :)

I also added a test, and a test file based on https://github.com/asraa/slsa-github-generator/blob/2811aeeec133e9841385f2d9c285d87d7c0150ed/.github/workflows/builder_docker-based_slsa3.yml#L183-L204; but I had to modify the JSON representation. Please let me know if the new json file makes senes. Without those changes unmarshaling it to `BuildDefinition` would not work.